### PR TITLE
Introduce GenericQuery API as a potential replacement for Query class

### PR DESCRIPTION
**Pull request by suresh-prakash**
## Description
1. Divide the expression types into 6 categories namely,
   * Constant
   * Literal
   * Arithmetic
   * Aggregate
   * Relational
   * Logical
2. Tag each expression type based on the supported clause(s) using the following marker interfaces
   Filterable: Relational, Logical
   Projectable: Constant, Literal, Arithmetic, Aggregate
   Groupable: Literal, Arithmetic
   Sortable: Literal, Arithmetic, Aggregate
3. Integrate these interfaces with necessary wrappers into the GenericQuery class, along with other fields
4. Each expression and the whole query can be constructed using builders or static constructors based on the
   number of fields

### Testing
Unit tests are added. Integration tests will be added in a separate PR.

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

### CODE REVIEW

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregateExpression.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Expression representing aggregation in a query.
+ *
+ * <p>Example: SUM(col1) can be constructed as <code>
+ *    AggregatorExpression.of(Aggregator.SUM, LiteralExpression.of("col1"));
+ * </code>
+ */
+@Value(staticConstructor = "of")
+public class AggregateExpression implements Projectable, Sortable {
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregateExpression.java:13]: Marker interfaces are often considered an antipattern (here's the first article I found talking about why: https://blog.ndepend.com/marker-interface-isnt-pattern-good-idea/ )- it seems like what we're really going for here is trying to get around the fact that java doesn't have a strong concept of unions. Either putting methods on the interfaces or using a different pattern (such as how oneof is represented in the java generated by protoc) would likely be a better fit.

Comment (aaron-steinfeld) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregateExpression.java:13]: Is it possible for something to be projectable but not sortable? Or vice versa? If not, it may make sense to merge the two.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/Aggregator.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package org.hypertrace.core.documentstore.expression;
+
+public enum Aggregator {
+  COUNT,
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/Aggregator.java:4]: distinct count is often more interesting than count and many of our use cases are built on it - probably want that too.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/ArithmeticExpression.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ *   ArithmeticExpression.builder()
+ *     .operator(ArithmeticOperator.SUBTRACT)
+ *     .operand(LiteralExpression.of("A"))
+ *     .operand(ConstantExpression.of(5))
+ *     .build();
+ * </code>
+ */
+@Value
+@Builder
+public class ArithmeticExpression implements Groupable, Projectable, Sortable {
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/ArithmeticExpression.java:28]: I'm not sure arithmetic is the right name here. This is really representing any non-aggregate expression as I read it (e.g. `substring(my_col, 0, 3)`)

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/ArithmeticOperator.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package org.hypertrace.core.documentstore.expression;
+
+public enum ArithmeticOperator {
+  // Unary operations
+  ABS,
+  FLOOR,
+
+  // Binary operations
+  SUBTRACT, // Minuend and then subtrahend
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/ArithmeticOperator.java:9]: both of these could be considered n-ary in the same way as add and multiply. e.g. `A - B - C` is just as valid as `A + B + C`

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/Filterable.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package org.hypertrace.core.documentstore.expression;
+
+/**
+ * A marker interface to represent that the expression is filterable and can be used in either the
+ * WHERE clause or the HAVING clause of the query.
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/Filterable.java:5]: where and having would be different, I believe. Having would need an aggregate expression, while where would need to be non-aggregated (I really need a better term for that).

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/Filterable.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+ */
+public interface Filterable {}
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/Filterable.java:7]: Independent of the other comments about marker interfaces and aggregates, I'm not sure these names are quite right. `Filterable`, in my mind, implies something can be filtered. Here, we're using this to actually say it does the filtering (maybe `Predicate` would be a better fit for this one? Same idea applies to others - more generically, could do things like `FilteringExpression`)

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/LiteralExpression.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+import lombok.Value;
+
+/**
+ * Expression representing either a literal (or a column name)
+ *
+ * <p>Example: LiteralExpression.of("col1");
+ */
+@Value(staticConstructor = "of")
+public class LiteralExpression implements Groupable, Projectable, Sortable {
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/LiteralExpression.java:11]: Can bring over the comment about IdentifyingExpression or ColumnExpression here.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/RelationalExpression.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+ *     company IN ('Traceable', 'Harness')
+ *  </code> can be constructed as <code>
+ *     RelationalExpression.of(
+ *         LiteralExpression.of("company"),
+ *         RelationalOperator.IN,
+ *         ConstantExpression.ofStrings("Traceable", "Harness"))));
+ * </code>
+ */
+@Value(staticConstructor = "of")
+public class RelationalExpression implements Filterable {
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/RelationalExpression.java:18]: would `company EXISTS` be a relational expression? If we don't need it, can omit - but we should ensure that adding it would not change the design.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/RelationalOperator.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+public enum RelationalOperator {
+  EQ,
+  NEQ,
+  GT,
+  LT,
+  GTE,
+  LTE,
+  IN,
+  CONTAINS,
+  EXISTS,
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/RelationalOperator.java:12]: It would be :) so above comment applies.

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericOrderBy.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+import lombok.Value;
+import org.hypertrace.core.documentstore.expression.Sortable;
+
+/**
+ * A generic ORDER BY definition that supports expressions. Note that this class is a more general
+ * version of {@link org.hypertrace.core.documentstore.OrderBy}
+ */
+@Value(staticConstructor = "of")
+public class GenericOrderBy {
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericOrderBy.java:11]: The use of the term "Generic" here seems strange - isn't everything in this tree generic? If the idea is to keep it distinct from that other class, I think the package name would be the appropriate place to make that distinction (and a corresponding comment similar to what's here).

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericOrderBy.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  Sortable sortable;
+  boolean ascending;
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericOrderBy.java:13]: personal nit: I've always found using a boolean to represent this weird, but may just be me. I just find something like `setDirection(DESC)` compared to `setAscending(false)` to be more readable.

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+ *                 false))
+ *         .offset(5)
+ *         .limit(10)
+ *         .build();
+ *  </code>
+ */
+@Value
+@Builder
+public class GenericQuery {
+  List<Projection> projections;
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java:63]: naming: is selections more generic?

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  Filterable whereFilter;
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java:64]: we should try to take as much SQL lingo out as we can without losing clarity, given this is an abstraction over some fairly different styles of database.

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+  @Singular List<Groupable> groupBys;
+  Filterable havingFilter;
+
+  @Singular List<GenericOrderBy> orderBys;
+
+  int offset;
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java:71]: I don't know what our existing support looks like - do we normally work in cursors, offsets or both?

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+  int limit;
+
+  public static class GenericQueryBuilder {
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java:74]: Why do we need our own builder - are there shortcomings in what we can represent in lombok?

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregateExpression.java [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+  Aggregator aggregator;
+  Projectable expression;
```
[the remark this answers is not in the document]
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregateExpression.java:15]: It would have to be translated to
`SELECT COUNT(1)`, which can then be substituted with `ConstantExpression.of(1)`

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregateExpression.java (hunk 1 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregateExpression.java:13]: This is very useful! Thank you. :) 
Read this article https://hackernoon.com/are-interfaces-code-smell-bd19abc266d3 and changed it to a visitor design pattern. The only difference here is that we have multiple hierarchies here instead of one.


On another note, ConstantExpressions are projectable and not sortable though SQL allows integers in "ORDER BY" clause. For example, ORDER BY 2, sorts by 2nd field in the projection list. We do not support that as of today as this implementation is not going to be tied to any specific SQL.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/ArithmeticExpression.java (hunk 3 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/ArithmeticExpression.java:28]: Renamed it as FunctionExpression as "SUBSTRING", "ADD", "SUBTRACT" are all functions. Would be happy to change it if anyone has a better alternative.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/Filterable.java (hunk 5 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/Filterable.java:5]: While that's true, MongoDB allows alias names to be used in "HAVING" clauses. So, I think, it's better to leave it this way since an invalid expression would anyway fail while querying the DB.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/Filterable.java (hunk 6 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/Filterable.java:7]: Makes sense! Thanks for the naming suggestions. :)

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericOrderBy.java (hunk 11 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericOrderBy.java:13]: Makes sense. I also prefer this better.

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java (hunk 13 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java:64]: Yup. Done some changes related to naming. Let me know if you find anything else specific to SQL.

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java (hunk 14 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java:71]: I see that the existing implementation exposes offsets outside (and works with cursors internally).

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java (hunk 15 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java:74]: This was done to avoid an unnecessary step while building the query. The implemented builder methods do the conversion from SelectingExpression to Selection internally, so that the clients can avoid the conversion step every single time.
`query.selection(Selection.of(LiteralExpression.of("col1")))` can be simplified as `query.selection(LiteralExpression.of("col1"))`.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/Aggregator.java (hunk 2 above)
Comment (varsha-abhinandan) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/Aggregator.java:4]: +1

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoFilteringExpressionParser.java [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+import org.hypertrace.core.documentstore.expression.LogicalExpression;
+import org.hypertrace.core.documentstore.expression.RelationalExpression;
+import org.hypertrace.core.documentstore.parser.IFilteringExpressionParser;
+
+public class MongoFilteringExpressionParser implements IFilteringExpressionParser {
+  private final static String FILTER_CLAUSE = "$match";
+
+  @Override
+  public Object parse(LogicalExpression expression) {
+    return null;
```
Comment (suresh-prakash) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoFilteringExpressionParser.java:14]: Have just implemented a sample section. Will implement the rest once the API is accepted.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoQueryParserTest.java [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.hypertrace.core.documentstore.expression.ConstantExpression;
+import org.hypertrace.core.documentstore.expression.FilteringExpression;
+import org.hypertrace.core.documentstore.expression.IdentifierExpression;
+import org.hypertrace.core.documentstore.expression.RelationalExpression;
+import org.hypertrace.core.documentstore.expression.RelationalOperator;
+import org.hypertrace.core.documentstore.query.Query;
+
+// TODO: Remove this testing class
+public class MongoQueryParserTest {
```
Comment (suresh-prakash) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoQueryParserTest.java:11]: Test class. Will be removed after necessary UTs are added.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoRelationalExpressionParser.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+    SelectingExpression expression2 = expression.getOperand2();
+
+    MongoSelectingExpressionParser selectionParser = new MongoSelectingExpressionParser();
+    String key = Objects.toString(expression1.parse(selectionParser));
+    Object value = expression2.parse(selectionParser);
+
+    return generateMap(key, value, operator);
+  }
+
+  private static Map<String, Object> generateMap(String key, Object value, RelationalOperator operator) {
```
Comment (suresh-prakash) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoRelationalExpressionParser.java:26]: This method is copied from the existing implementation

File: document-store/src/main/java/org/hypertrace/core/documentstore/parser/IFilteringExpressionParser.java [hunk 20]
```diff
@@ -0,0 +1,10 @@
+package org.hypertrace.core.documentstore.parser;
+
+import org.hypertrace.core.documentstore.expression.LogicalExpression;
+import org.hypertrace.core.documentstore.expression.RelationalExpression;
+
+public interface IFilteringExpressionParser {
```
Comment (varsha-abhinandan) [inline document-store/src/main/java/org/hypertrace/core/documentstore/parser/IFilteringExpressionParser.java:6]: It's a personal preference but I prefer not using the prefix "I" for interfaces since interfaces are what gets used widely and it hurts readability. We could just simply use FilteringExpressionParser since the impl class anyway has the db type prefix like MongoFilteringExpressionParser

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregationOperator.java [hunk 21]
```diff
@@ -0,0 +1,8 @@
+package org.hypertrace.core.documentstore.expression;
+
+public enum AggregationOperator {
```
Comment (varsha-abhinandan) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregationOperator.java:3]: can we move all the various operators under a package of its own and similarly the interfaces and the impl classes

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+ *                     .operand(IdentifierExpression.of("col8")).build(),
+ *                 DESC))
+ *         .paginationDefinition(PaginationDefinition.of(5, 10))
+ *         .build();
+ *  </code>
+ */
+@Value
+@Builder
+public class Query {
+  List<Selection> selections;
```
Comment (varsha-abhinandan) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:62]: Can't we use SelectionExpression here similar to FilteringExpression , GroupingExpression?

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/FilteringExpression.java [hunk 23]
```diff
@@ -0,0 +1,11 @@
+package org.hypertrace.core.documentstore.expression;
+
+import org.hypertrace.core.documentstore.parser.FilteringExpressionParser;
+
+/**
+ * An interface to represent that the expression can be used in either the WHERE clause or the
```
Comment (varsha-abhinandan) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/FilteringExpression.java:6]: It might be cleaner to separate out the where and having expressions. Having clause is for aggregates. So, separating these two will make it less error-prone

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java (hunk 22 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:62]: Selection is a combination of SelectionExpression and an alias. So, SelectionExpression can't be directly used here.

Comment (varsha-abhinandan) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:62]: If an alias is required for all selections, should we have this in the abstract class and have the impls of  SelectionExpression extend it?

Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:62]: Even I thought of doing the same. However, there are 2 reasons why it is not done that way.

1. Not all subclasses of SelectingExpression will need an "alias" field, because a subclass of SelectingExpression might actually be used as a SortingExpression or vice versa. Here, inheritance is purely used for type-substitution. Introducing "alias" to SelectingExpression as a class will end up in not only unwanted memory, but, also change the way a Query is constructed (in an undesired manner).
2. Similar to SelectingExpression, SortingExpression can also be converted to an abstract class with the "SortingOrder" field. When we do so, classes like "AggregateExpression" will need multiple inheritance, which is not supported in Java.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/FilteringExpression.java (hunk 23 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/FilteringExpression.java:6]: Even Aaron had given the same comment. Ref.: https://github.com/hypertrace/document-store/pull/58#discussion_r745134184.

1. Even after separating out HavingExpression separately, it would have an identical structure since databases like MongoDB support aliases (IdentifierExpression) to be used in the conditional statements (than just AggregateExpressions)
2. For databases with different HAVING and WHERE implementations (like Postgres), it can be separated out at the Parser hierarchy.

In summary, a filter is common despite its usage in WHERE/HAVING clauses. The differentiation can be done in the implementation (at the Parser hierarchy).

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoFunctionExpressionParser.java [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.stream.Collectors;
+import org.hypertrace.core.documentstore.expression.impl.FunctionExpression;
+import org.hypertrace.core.documentstore.parser.SelectingExpressionParser;
+
+public class MongoFunctionExpressionParser {
+
+  static Map<String, Object> parse(FunctionExpression expression) {
+    int numArgs = expression.getOperands().size();
+
+    if (numArgs == 0) {
```
Comment (suresh-prakash) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoFunctionExpressionParser.java:14]: Probably these validations are no longer required with Hibernate validations taking place while building the query.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoSelectingExpressionParser.java [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+import org.hypertrace.core.documentstore.expression.impl.ConstantExpression;
+import org.hypertrace.core.documentstore.expression.impl.FunctionExpression;
+import org.hypertrace.core.documentstore.expression.impl.IdentifierExpression;
+import org.hypertrace.core.documentstore.parser.SelectingExpressionParser;
+
+public class MongoSelectingExpressionParser implements SelectingExpressionParser {
+
+  @Override
+  public Object parse(AggregateExpression expression) {
+    return null;
```
Comment (suresh-prakash) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoSelectingExpressionParser.java:13]: TODO

**Comment by saxenakshitiz** [comment]
Overall lgtm

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Selection.java [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+ * {@link IdentifierExpression}, the alias is inferred if not supplied.
+ */
+@Value
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class Selection {
+
+  // Special selection representing the selection of all the columns
+  public static final Selection ALL = new Selection(null, null);
+
+  @NotNull SelectingExpression expression; // Can be null only in the case of select all columns
```
[the remark this answers is not in the document]
Comment (varsha-abhinandan) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Selection.java:23]: As discussed, let's model ALL as well as a selection expression instead of dealing with nulls.

[the remark this answers is not in the document]
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Selection.java:23]: Sure

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+  @Size(min = 1)
+  List<@NotNull SortingDefinition> sortingDefinitions;
+
+  // Missing pagination definition represents fetching all the records
+  PaginationDefinition paginationDefinition;
+
+  public static class QueryBuilder {
+
+    public Query build() {
+      fillMissingSelections();
```
[the remark this answers is not in the document]
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:93]: Yeah. That makes perfect sense.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoAggregateExpressionParser.java [hunk 28]
```diff
@@ -0,0 +1,17 @@
+package org.hypertrace.core.documentstore.mongo.parser;
+
+import java.util.Map;
+import org.hypertrace.core.documentstore.expression.impl.AggregateExpression;
+import org.hypertrace.core.documentstore.parser.SelectingExpressionParser;
+
+public class MongoAggregateExpressionParser {
+
+  static Map<String, Object> parse(final AggregateExpression expression) {
+    String key = "$" + expression.getAggregator().name().toLowerCase();
```
Comment (varsha-abhinandan) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoAggregateExpressionParser.java:10]: nit: can we extract this out to a contant

**Comment by varsha-abhinandan** [comment]
@suresh-prakash Rest of the skeletal LGTM

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/AggregationOperator.java [hunk 29]
```diff
@@ -0,0 +1,8 @@
+package org.hypertrace.core.documentstore.expression.operators;
+
+public enum AggregationOperator {
```
Comment (varsha-abhinandan) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/AggregationOperator.java:3]: min, max

Comment (varsha-abhinandan) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/AggregationOperator.java:3]: though may not be needed currently

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+
+  FilteringExpression filter;
+
+  @Singular
+  @Size(min = 1)
+  List<@NotNull GroupingExpression> aggregations;
+
+  FilteringExpression aggregationFilter;
+
+  @Size(min = 1)
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:84]: Does this mean order by being mandatory in a request or will have a default?

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/WhitelistedSelection.java [hunk 31]
```diff
@@ -0,0 +1,42 @@
...
+    return expression instanceof AggregateExpression;
+  }
+
+  @Override
+  public boolean allColumnsSelected() {
+    return false;
+  }
+
+  public static Selection of(final SelectingExpression expression) {
+    return WhitelistedSelection.of(expression, null);
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/WhitelistedSelection.java:36]: why does validation not required in this method? Is validation is only for `alias`?

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java [hunk 32]
```diff
@@ -0,0 +1,135 @@
...
+@Builder
+public class Query {
+
+  @NotEmpty List<@NotNull Selection> selections;
+
+  FilteringExpression filter;
+
+  @Singular
+  @Size(min = 1)
+  List<@NotNull GroupingExpression> aggregations;
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:80]: I see that for `Selection` we have `definition and `Expression` objects separate. But, why are we not having the same pattern for `GroupBy` and `Filter`?

File: document-store/src/main/java/org/hypertrace/core/documentstore/parser/SelectingExpressionParser.java [hunk 33]
```diff
@@ -0,0 +1,16 @@
+package org.hypertrace.core.documentstore.parser;
+
+import org.hypertrace.core.documentstore.expression.impl.AggregateExpression;
+import org.hypertrace.core.documentstore.expression.impl.ConstantExpression;
+import org.hypertrace.core.documentstore.expression.impl.FunctionExpression;
+import org.hypertrace.core.documentstore.expression.impl.IdentifierExpression;
+
+public interface SelectingExpressionParser {
+  Object parse(AggregateExpression expression);
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/parser/SelectingExpressionParser.java:9]: Like `SortingExpressionParser`, shouldn't the rest of the interfaces of parse be taking `final` argument?

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutor.java [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.bson.conversions.Bson;
+import org.hypertrace.core.documentstore.query.PaginationDefinition;
+import org.hypertrace.core.documentstore.query.Query;
+
+@Slf4j
+public class MongoQueryExecutor {
+
+  public static Iterator<BasicDBObject> find(
+      final Query query, final com.mongodb.client.MongoCollection<BasicDBObject> collection) {
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutor.java:32]: Can we make these methods `non-static` and `collection will be its instance as mostly they will be used from the implementation of `Collection` interface?

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java (hunk 30 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:84]: Actually, it meant this value can be NULL or a list with at least one element (but, never an empty list).
However, after going through this in our discussion, even having an empty list will not harm as it is handled.
So, will remove this annotation from here.

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/WhitelistedSelection.java (hunk 31 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/WhitelistedSelection.java:36]: Validation is required here. The invoked method `WhitelistedSelection.of()` does the validation internally.
Validation is only for the SelectingExpression. "alias" doesn't require any validation as it can be null/blank.

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java (hunk 32 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:80]: Yeah. I see the issue in readability. Will consider this to improve readability in subsequent PRs (along with writing integration tests).

Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:80]: Keeping this conversation unresolved to address in a future PR.

File: document-store/src/main/java/org/hypertrace/core/documentstore/parser/SelectingExpressionParser.java (hunk 33 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/parser/SelectingExpressionParser.java:9]: Yup. Good catch. I added the missing `final`s

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/impl/RelationalExpression.java [hunk 35]
```diff
@@ -0,0 +1,45 @@
...
+ * </code> can be constructed as <code>
+ * RelationalExpression.of( IdentifierExpression.of("company"), RelationalOperator.IN,
+ * ConstantExpression.ofStrings("Traceable", "Harness"))));
+ * </code>
+ */
+@Value
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class RelationalExpression implements FilteringExpression {
+
+  @NotNull SelectingExpression operand1;
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/impl/RelationalExpression.java:28]: nit: I think, operand1 to lhs and operand2 to rhs would have been more readable

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutor.java [hunk 36]
```diff
@@ -0,0 +1,104 @@
...
+    BasicDBObject skipClause = getSkipClause(query.getPaginationDefinition());
+    BasicDBObject limitClause = getLimitClause(query.getPaginationDefinition());
+
+    BasicDBObject projectClause = getProjectClause(query.getSelections());
+
+    List<BasicDBObject> pipeline =
+        Stream.of(
+                filterClause,
+                groupClause,
+                groupFilterClause,
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutor.java:70]: Order will not matter, right?
Should `groupFilterClause` come before `groupClause`?

Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutor.java:70]: Yes. The order of specification is immaterial

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/impl/RelationalExpression.java (hunk 35 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/impl/RelationalExpression.java:28]: Makes sense. Will address in the follow-up PR

**Review by kotharironak — APPROVED** [review]
Overall LGTM. I will try out a few things once we have an integration test to play with as follow-up PR. See if any open comments from others.

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/AllSelection.java [hunk 37]
```diff
@@ -0,0 +1,20 @@
+package org.hypertrace.core.documentstore.query;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class AllSelection implements Selection {
```
Comment (varsha-abhinandan) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/AllSelection.java:7]: Thinking about this again - we probably don't need to support "all". Expecting specific fields should be good enough.

Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/AllSelection.java:7]: Yeah. With respect to MongoDB, selecting specific-fields is sufficient. Since Postgres (which is an SQL) is also supported as of now and for any future use cases, I think, it is good to have.

Since these APIs are not bound to a specific database, I wanted to keep it as generic as possible so that we don't break the compatibility frequently.

What do you think?

Comment (varsha-abhinandan) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/AllSelection.java:7]: Yeah, this isn't specific to a particular DB. It's whether our query interface needs to support "all" fields or not. The existing query interface also expects the user to specify the required fields upfront. In the new Query class, the top fields seem to have a mix of expressions, definitions, and now the selection interface. Was thinking if we can at least simplify the selection one by removing the * support and look into the other two. We can merge this PR and take the open comments from Ronak as well in your next.

**Review by varsha-abhinandan — APPROVED** [review]

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/AllSelection.java (hunk 37 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/AllSelection.java:7]: Sure. Will work on it in the next PR.
